File: src/data/capsuleStudio.ts
export type CapsuleRuntime = {
  id: string;
  name: string;
  purpose: string;
  languages: string[];
  target: "web-worker" | "wasm" | "wasi" | "node" | "native" | "hybrid";
  deployAs: "agent" | "app" | "worker" | "service";
  status: "ready" | "planned" | "toolchain-required";
};

export type CapsuleWorker = {
  name: string;
  type: "persistent" | "session" | "build" | "preview" | "deploy" | "research";
  input: string;
  output: string;
  status: "online" | "standby" | "planned";
};

export type CapsuleDeployTarget = {
  name: string;
  route: string;
  purpose: string;
  status: "ready" | "planned";
};

export type CapsuleDivision = {
  name: string;
  role: string;
  workers: string[];
};

export const capsuleRuntimes: CapsuleRuntime[] = [
  {
    id: "python-orchestrator",
    name: "Python Orchestrator Capsule",
    purpose: "Coordinates scripts, reports, artifact generation, local services, and AI tool calls.",
    languages: ["Python", "Shell", "JSON", "Markdown"],
    target: "wasi",
    deployAs: "agent",
    status: "ready"
  },
  {
    id: "julia-sciml",
    name: "Julia SciML Capsule",
    purpose: "Runs scientific modeling, numerical workloads, neural ODE planning, and research calculations.",
    languages: ["Julia", "Python"],
    target: "native",
    deployAs: "service",
    status: "toolchain-required"
  },
  {
    id: "cpp-wasm-core",
    name: "C/C++ WASM Core Capsule",
    purpose: "Compiles performance-critical scoring, export, simulation, and runtime kernels into WebAssembly.",
    languages: ["C", "C++", "WAT"],
    target: "wasm",
    deployAs: "worker",
    status: "ready"
  },
  {
    id: "react-app-capsule",
    name: "React App Capsule",
    purpose: "Packages full front-end apps with preview, project workspace, marketplace, and deploy flow.",
    languages: ["TypeScript", "React", "CSS", "HTML"],
    target: "web-worker",
    deployAs: "app",
    status: "ready"
  },
  {
    id: "node-service-capsule",
    name: "Node Service Capsule",
    purpose: "Runs API adapters, MCP bridges, preview routing, GitHub publishing, and background daemons.",
    languages: ["Node.js", "TypeScript", "JSON-RPC"],
    target: "node",
    deployAs: "service",
    status: "planned"
  },
  {
    id: "java-agent-capsule",
    name: "Java Agent Capsule",
    purpose: "Packages JVM services and agent runtimes into managed capsule builds.",
    languages: ["Java", "Kotlin"],
    target: "wasm",
    deployAs: "agent",
    status: "planned"
  }
];

export const capsuleWorkers: CapsuleWorker[] = [
  { name: "Capsule Session Worker", type: "session", input: "selected language + source files", output: "live session manifest", status: "online" },
  { name: "Preview Worker", type: "preview", input: "web app, API, terminal, notebook, or artifact", output: "clickable preview surface", status: "online" },
  { name: "WASM Build Worker", type: "build", input: "C, C++, Rust, Go, Java targets", output: "wasm or wasi capsule module", status: "standby" },
  { name: "Agent Deploy Worker", type: "deploy", input: "capsule manifest + artifacts", output: "agent, worker, service, or app deployment", status: "standby" },
  { name: "Research Report Worker", type: "research", input: "capsule run logs + model outputs", output: "launch, R&D, architecture, and verification reports", status: "online" },
  { name: "GitHub Publisher Worker", type: "deploy", input: "generated source tree", output: "repo commit / handoff", status: "online" },
  { name: "Persistent Background Daemon", type: "persistent", input: "workspace events", output: "auto-refresh, indexing, and status updates", status: "planned" }
];

export const capsuleDeployTargets: CapsuleDeployTarget[] = [
  { name: "Web Worker Agent", route: "workers/<capsule>.worker.ts", purpose: "Run capsule logic inside the browser without blocking the UI.", status: "ready" },
  { name: "Whole Web App", route: "apps/<capsule>/", purpose: "Deploy a full HTML/React/Vite app capsule with preview and publish metadata.", status: "ready" },
  { name: "WASM Kernel", route: "wasm/<capsule>.wasm", purpose: "Ship compiled kernels for scoring, simulation, parsing, and model support.", status: "ready" },
  { name: "MCP Service", route: "mcp/<capsule>/server", purpose: "Expose capsule tools to AI clients through tool protocols.", status: "planned" },
  { name: "GitHub Release", route: "github://ItsNotAILABS/specforge-launch-studio", purpose: "Commit source, docs, manifests, reports, and deployment instructions.", status: "ready" }
];

export const capsuleDivisions: CapsuleDivision[] = [
  { name: "Chief Orchestration", role: "Coordinates Python, Julia, Node, WASM, GitHub, and worker deployment lanes.", workers: ["Capsule Session Worker", "Agent Deploy Worker", "GitHub Publisher Worker"] },
  { name: "Chief Research", role: "Generates R&D reports, architecture briefs, model plans, and verification packages.", workers: ["Research Report Worker", "Persistent Background Daemon"] },
  { name: "Runtime Engineering", role: "Builds web workers, WASM kernels, preview routes, and capsule manifests.", workers: ["WASM Build Worker", "Preview Worker"] },
  { name: "Launch Operations", role: "Turns capsules into marketable apps, services, or deployable agents.", workers: ["Agent Deploy Worker", "GitHub Publisher Worker"] }
];

File: tools/capsule_studio.py
#!/usr/bin/env python3
"""Capsule Studio local manifest and report utility.

This script does not pretend to compile every language. It creates execution-ready
capsule manifests and verifies the capsule metadata that local runtimes can use.
"""

from __future__ import annotations

import argparse
import json
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import List

ROOT = Path(__file__).resolve().parents[1]
CAPSULE_DIR = ROOT / "capsules"
EXAMPLE_DIR = CAPSULE_DIR / "examples"
ARTIFACT_DIR = ROOT / "artifacts" / "capsule-studio"


@dataclass
class CapsuleRuntime:
    id: str
    name: str
    languages: List[str]
    targets: List[str]
    deploy_mode: str
    destination: str
    server: bool
    preview: str


RUNTIMES = {
    "python": CapsuleRuntime(
        id="python-orchestrator",
        name="Python Orchestrator Capsule",
        languages=["Python", "JSON", "Markdown", "Shell"],
        targets=["wasi", "native", "github"],
        deploy_mode="agent",
        destination="github://ItsNotAILABS/specforge-launch-studio",
        server=True,
        preview="terminal",
    ),
    "react": CapsuleRuntime(
        id="react-worker-agent",
        name="React Worker Agent Capsule",
        languages=["TypeScript", "React", "CSS", "HTML"],
        targets=["browser", "web-worker", "github"],
        deploy_mode="app",
        destination="github://ItsNotAILABS/specforge-launch-studio",
        server=True,
        preview="web",
    ),
    "cpp": CapsuleRuntime(
        id="cpp-wasm-core",
        name="C++ WASM Core Capsule",
        languages=["C", "C++", "WAT"],
        targets=["wasm", "web-worker", "github"],
        deploy_mode="web-worker",
        destination="workers/specforge_core.worker",
        server=False,
        preview="terminal",
    ),
    "node": CapsuleRuntime(
        id="node-service-capsule",
        name="Node Service Capsule",
        languages=["Node.js", "TypeScript", "JSON-RPC"],
        targets=["node", "github"],
        deploy_mode="service",
        destination="mcp/node-service",
        server=True,
        preview="api",
    ),
    "julia": CapsuleRuntime(
        id="julia-sciml-capsule",
        name="Julia SciML Capsule",
        languages=["Julia", "Python", "Markdown"],
        targets=["native", "github"],
        deploy_mode="service",
        destination="runtimes/julia-sciml",
        server=True,
        preview="notebook",
    ),
}


def build_manifest(runtime: CapsuleRuntime) -> dict:
    return {
        "id": runtime.id,
        "name": runtime.name,
        "version": "0.1.0",
        "runtime": {
            "kind": runtime.id.split("-")[0],
            "entrypoint": "defined-by-capsule",
            "server": runtime.server,
            "preview": runtime.preview,
        },
        "languages": runtime.languages,
        "targets": runtime.targets,
        "workers": [
            {"name": "Capsule Session Worker", "role": "session manifest and lifecycle", "persistent": True},
            {"name": "Preview Worker", "role": "web api terminal notebook artifact preview", "persistent": False},
            {"name": "GitHub Publisher Worker", "role": "repo handoff and deploy metadata", "persistent": False},
        ],
        "deploy": {"mode": runtime.deploy_mode, "destination": runtime.destination},
    }


def write_manifest(runtime_key: str) -> Path:
    if runtime_key not in RUNTIMES:
      raise SystemExit(f"unknown runtime: {runtime_key}")
    ARTIFACT_DIR.mkdir(parents=True, exist_ok=True)
    manifest = build_manifest(RUNTIMES[runtime_key])
    path = ARTIFACT_DIR / f"{manifest['id']}.capsule.json"
    path.write_text(json.dumps(manifest, indent=2) + "\n", encoding="utf-8")
    return path


def verify_examples() -> int:
    if not EXAMPLE_DIR.exists():
        print("no capsule examples found")
        return 1
    failures = 0
    for path in sorted(EXAMPLE_DIR.glob("*.capsule.json")):
        data = json.loads(path.read_text(encoding="utf-8"))
        required = ["id", "name", "version", "runtime", "languages", "targets", "workers", "deploy"]
        missing = [key for key in required if key not in data]
        if missing:
            failures += 1
            print(f"FAIL {path}: missing {missing}")
        else:
            print(f"OK {path.name}: {data['deploy']['mode']} -> {data['deploy']['destination']}")
    return failures


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    sub = parser.add_subparsers(dest="command", required=True)
    sub.add_parser("list")
    make = sub.add_parser("make")
    make.add_argument("runtime", choices=sorted(RUNTIMES))
    sub.add_parser("verify")
    args = parser.parse_args()

    if args.command == "list":
        print(json.dumps({key: asdict(value) for key, value in RUNTIMES.items()}, indent=2))
        return 0
    if args.command == "make":
        path = write_manifest(args.runtime)
        print(path)
        return 0
    if args.command == "verify":
        return verify_examples()
    return 1


if __name__ == "__main__":
    raise SystemExit(main())
